humanoid_league_team_communication/src/humanoid_league_team_communication/humanoid_league_team_communication.py

```diff
@@ -0,0 +1,398 @@
+
+            if ball.confidence == 0:
+                self.ball = None
+            else:
+                # Transform to map
+                ball_point = PointStamped(msg.header, ball.pose.pose.position)
+                try:
+                    self.ball = self.tf_buffer.transform(ball_point, "map", timeout=rospy.Duration.from_sec(0.3))
+                    self.ball_confidence = ball.confidence
+                except tf2_ros.TransformException:
+                    self.ball = None
+
+    def __del__(self):
+        self.close_connection()
+
+    def receive_forever(self):
+        while not rospy.is_shutdown():
+            try:
+                msg = self.receive_msg()
+            except (struct.error, socket.timeout):
+                continue
+
+            if msg:  # Not handle empty messages or None
+                self.handle_message(msg)
+
+    def handle_message(self, msg):
+        def set_covariance(fmat3, ros_covariance):
+            # ROS covariance is row-major 36 x float, protobuf covariance is column-major 9 x float [x, y, θ]
+            ros_covariance[0] = fmat3.x.x
+            ros_covariance[1] = fmat3.y.x
+            ros_covariance[5] = fmat3.z.x
+            ros_covariance[6] = fmat3.x.y
+            ros_covariance[7] = fmat3.y.y
+            ros_covariance[11] = fmat3.z.y
+            ros_covariance[30] = fmat3.x.z
+            ros_covariance[31] = fmat3.y.z
+            ros_covariance[35] = fmat3.z.z
+
+        def set_pose(robot, pose):
+            pose.pose.position.x = robot.position.x
+            pose.pose.position.y = robot.position.y
+
+            quat = transforms3d.euler.euler2quat(0.0, 0.0, robot.position.z)  #wxyz -> ros: xyzw
+            pose.pose.orientation.x = quat[1]
+            pose.pose.orientation.y = quat[2]
+            pose.pose.orientation.z = quat[3]
+            pose.pose.orientation.w = quat[0]
+
+            if pose.covariance:
+                set_covariance(robot.covariance, pose.covariance)
+
+        message = robocup_extension_pb2.Message()
+        message.ParseFromString(msg)
+
+        player_id = message.current_pose.player_id
+        team_id = message.current_pose.team
+
+        if player_id == self.player_id or team_id != self.team_id:
+            # Skip information from ourselves or from the other team
+            return
+
+        team_data = TeamData()
+
+        header = Header()
+        header.stamp.secs = message.timestamp.seconds
+        header.stamp.nsecs = message.timestamp.nanos
+        header.frame_id = "map"
+
+        # Handle timestamp
+        ##################
+        team_data.header = header
+
+        # Handle robot ID
+        #################
+        team_data.robot_id = player_id
+
+        # Handle state
+        ##############
+        team_data.state = message.state
+
+        # Handle pose of current player
+        ###############################
+        set_pose(message.current_pose, team_data.robot_position.pose)
```

Comment:
I find this function without a return rather confusing in the python context

Reply:
The problem is that in Protobuf, you cannot just create another object and simply assign it in the message, you can only assign primitive types. But maybe CopyFrom would be possible to use here, I definitely agree that what is done here is not the pythonic way